src/main/java/org/passay/entropy/ShannonEntropyFactory.java

```diff
@@ -38,57 +56,26 @@ public static ShannonEntropy createEntropy(final List<Rule> passwordRules, final
     if (!passwordData.getOrigin().equals(PasswordData.Origin.User)) {
       throw new IllegalArgumentException("Password data must have an origin of " + PasswordData.Origin.User);
     }
-    boolean dictionaryCheck = false;
-    boolean compositionCheck = false;
-    for (Rule rule : passwordRules) {
-      if (!compositionCheck) {
-        if (rule instanceof CharacterCharacteristicsRule) {
-          final CharacterCharacteristicsRule characteristicRule = (CharacterCharacteristicsRule) rule;
-          for (CharacterRule r : characteristicRule.getRules()) {
-            compositionCheck = hasComposition(r, r.getValidCharacters(), passwordData);
-            if (compositionCheck) {
-              break;
-            }
-          }
-        } else if (rule instanceof CharacterRule) {
-          final CharacterRule characterRule = (CharacterRule) rule;
-          compositionCheck = hasComposition(characterRule, characterRule.getValidCharacters(), passwordData);
-        } else if (rule instanceof AllowedCharacterRule) {
-          final AllowedCharacterRule allowedCharacterRule = (AllowedCharacterRule) rule;
-          compositionCheck = hasComposition(
-            rule,
-            String.valueOf(allowedCharacterRule.getAllowedCharacters()),
-            passwordData);
-        }
-      }
-      if (AbstractDictionaryRule.class.isAssignableFrom(rule.getClass())) {
-        dictionaryCheck = true;
-      }
-    }
+    final boolean dictionaryCheck = passwordRules.stream().filter(
+      rule -> AbstractDictionaryRule.class.isAssignableFrom(
+        rule.getClass()) && ((AbstractDictionaryRule) rule).getDictionary().size() > 0).count() > 0;
+    final boolean compositionCheck = hasComposition(passwordData);
     return new ShannonEntropy(dictionaryCheck, compositionCheck, passwordData.getPassword().length());
   }
 
 
   /**
-   * Determine if the supplied rule and characters has composition. (As defined by NIST SP-800-63-1 a password has
-   * composition rules if it checks for anything other than lowercase letter password).
+   * Checks whether the supplied passwordData has composition. (As suggested by NIST SP-800-63-1)
    *
-   * @param  rule  to validate for the matching composition rule
-   * @param  characters  to check for composition
-   * @param  passwordData  with the contents of the password.
+   * @param  passwordData  to check for composition
    *
-   * @return  whether the supplied rule has the correct composition
+   * @return  true if valid, false otherwise
    */
-  private static boolean hasComposition(final Rule rule, final String characters, final PasswordData passwordData)
+  protected static boolean hasComposition(final PasswordData passwordData)
```

Comment:
Good catch, I actually had a class I did not end up including the commit, we can revert that to private.